apps/emqx_connector_aggregator/src/emqx_connector_aggreg_delivery.erl

```diff
@@ -0,0 +1,183 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022-2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+%% This module takes aggregated records from a buffer and delivers them to a blob storage
+%% backend, wrapped in a configurable container (though currently there's only CSV).
+-module(emqx_connector_aggreg_delivery).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+-include("emqx_connector_aggregator.hrl").
+
+-export([start_link/3]).
+
+%% Internal exports
+-export([
+    init/4,
+    loop/3
+]).
+
+%% Sys
+-export([
+    system_continue/3,
+    system_terminate/4,
+    format_status/2
+]).
+
+-record(delivery, {
+    name :: _Name,
+    callback_module :: module(),
+    container :: emqx_connector_aggreg_csv:container(),
+    reader :: emqx_connector_aggreg_buffer:reader(),
+    upload :: impl_specific_upload_state(),
+    empty :: boolean()
+}).
+
+-type state() :: #delivery{}.
+
+-type init_opts() :: #{
+    callback_module := module(),
+    any() => term()
+}.
+
+-type impl_specific_upload_state() :: term().
+
+-callback init_upload_state(buffer(), map()) -> impl_specific_upload_state().
+
+-callback process_append(iodata(), impl_specific_upload_state()) ->
+    {ok, impl_specific_upload_state()}.
+
+-callback process_write(impl_specific_upload_state()) -> impl_specific_upload_state().
```

Comment:
From the generalization standpoint, feels a bit weird: I'd expect that `append/2` does not fail (because it basically implies no IO) but `write/1` does.